### PR TITLE
Add optimizer_for_mobile doc into python api root doc

### PR DESCRIPTION
**Pull request by xcheng16**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41211 Add optimizer_for_mobile doc into python api root doc**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22543608](https://our.internmc.facebook.com/intern/diff/D22543608)

### CODE REVIEW

File: docs/source/mobile_optimizer.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+Torch mobile supports :mod:`torch.mobile_optimizer.optimize_for_mobile` utility to run a list of optimization pass with modules in eval mode.
+The method takes the following parameters:
+
+1. A TorchScript object;
+2. A set of blacklisted optimization;
+3. A list of preserved methods.
+
+
+.. note::
+By default, if optimization blacklist is None or empty, “optimize_for_mobile” will run the following optimizations:
```
Comment (jlin27) [inline docs/source/mobile_optimizer.rst:15]: Suggestion: Use either italics or code syntax for optimize_for_mobile of quotation marks.

File: docs/source/mobile_optimizer.rst [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+3. ReLU/Hardtanh fusion: XNNPACK ops support fusion of clamping. That is clamping of output activation is done as part of the kernel, including for 2D convolution and linear op kernels. Thus clamping effectively comes for free. Thus any op that can be expressed as clamping op, such as ReLU or hardtanh, can be fused with previous Conv2D or linear op in XNNPACK. This pass rewrites graph by finding ReLU/hardtanh ops that follow XNNPACK Conv2D/linear ops, written by the previous pass, and fuses them together.
+4. MobileOptimizerType::REMOVE_DROPOUT: This optimization pass removes dropout and dropout_ nodes from this module when training is false.
+
+
+“optimize_for_mobile” will also invoke freeze_module pass which only preserve only forward method. If you have other method to that needed to be preserved,  add them into the preserved method list and pass into the method.
```
Comment (jlin27) [inline docs/source/mobile_optimizer.rst:23]: Suggestion: Use either italics or code syntax for optimize_for_mobile of quotation marks.

File: docs/source/mobile_optimizer.rst [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+1. A TorchScript object;
+2. A set of blacklisted optimization;
+3. A list of preserved methods.
+
+
+.. note::
+By default, if optimization blacklist is None or empty, “optimize_for_mobile” will run the following optimizations:
+
+1. MobileOptimizerType::CONV_BN_FUSION:  This optimization pass folds Conv2d-BatchNorm2d into Conv2d in forward method of this module and all its submodules. The weight and bias of the Conv2d are correspondingly updated.
```
Comment (jlin27) [inline docs/source/mobile_optimizer.rst:17]: Suggestion: Link  or bold so it is easier to see. For example:
```
**MobileOptimizerType::CONV_BN_FUSION:**
```

File: docs/source/mobile_optimizer.rst [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+3. ReLU/Hardtanh fusion: XNNPACK ops support fusion of clamping. That is clamping of output activation is done as part of the kernel, including for 2D convolution and linear op kernels. Thus clamping effectively comes for free. Thus any op that can be expressed as clamping op, such as ReLU or hardtanh, can be fused with previous Conv2D or linear op in XNNPACK. This pass rewrites graph by finding ReLU/hardtanh ops that follow XNNPACK Conv2D/linear ops, written by the previous pass, and fuses them together.
+4. MobileOptimizerType::REMOVE_DROPOUT: This optimization pass removes dropout and dropout_ nodes from this module when training is false.
+
+
+“optimize_for_mobile” will also invoke freeze_module pass which only preserve only forward method. If you have other method to that needed to be preserved,  add them into the preserved method list and pass into the method.
+
+.. currentmodule:: torch.utils.mobile_optimizer
```
Comment (jlin27) [inline docs/source/mobile_optimizer.rst:25]: This function isn't rendering for me. Are you seeing it when you build locally?

**Comment by jlin27** [comment]
Hi @xcheng16 - I just have a few suggestions for formatting and code rendering. Here's how it looks when I build locally:  screencapture-file-Users-jplin-github-pytorch-docs-build-html-mobile-optimizer-html-2020-07-09-16_57_46

File: docs/source/mobile_optimizer.rst [hunk 5]
```diff
@@ -0,0 +1,26 @@
+torch.utils.mobile_optimizer
```
[the remark this answers is not in the document]
Comment (xcheng16) [reply docs/source/mobile_optimizer.rst:1]: sure, I have this in the local repo, will update it.

**Comment by xcheng16** [comment]
@jlin27 Below is latest version on how it looks like locally.
Screen Shot 2020-07-10 at 12 34 44 PM

**Comment by kimishpatel** [comment]
> @jlin27 Below is latest version on how it looks like locally.
> Screen Shot 2020-07-10 at 12 34 44 PM

Maybe instead of naming pass with MobileOptimizerType::CONV_BN_FUSION, we should just put in bold the name of the pass like how ReLU/Hardtanh fusion is. Otherwise there is a discrepancy between how ReLU/Hardtanh looks vs rest. I guess we have not added an option to disable relu fusion.

It may look like:
```
- **Conv2D + BatchNorm fusion** (blacklisting option `MobileOptimizerType::CONV_BN_FUSION`:  This optimization pass folds ``Conv2d-BatchNorm2d`` into ``Conv2d`` in ``forward`` method of this module and all its submodules. The weight and bias of the ``Conv2d`` are correspondingly updated.
```
Not sure if this is better. But just a suggestion.

**Comment by xcheng16** [comment]
@kimishpatel I like this idea.

**Review by jlin27 — APPROVED** [review]

**Pull request closed without merging**